---
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r include = F}
library(dplyr)
library(data.table)
library(gdsfmt)
library(RColorBrewer)

source("~/kul/dex-stim-human-array-isns/code/00_functions/load_gds_files.R")
source("~/kul/dex-stim-human-array-isns/code/00_functions/get_chord_diagramm.R")

cbPalette <- c("grey", "#0072B2")
brewer_pallete <- brewer.pal(n = 8, name = "Dark2")
```

```{r include = F}
dir_pre  <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/"
rslt_dir <- "/binder/mgp/workspace/2020_DexStim_Array_Human/dex-stim-human-isns/results/5_fold_cv/all/"
setwd(rslt_dir)
```

```{r load-anno, include = F}
treatment       <- "dex"
dex_cpg_anno_gr <- readRDS(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_cpgs_annotated.rds"))
dex_snp_anno_gr <- readRDS(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_snps_annotated.rds"))


treatment       <- "veh"
veh_cpg_anno_gr <- readRDS(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_cpgs_annotated.rds"))
veh_snp_anno_gr <- readRDS(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_snps_annotated.rds"))

# loc_df <- rbind(dex_loc_df, veh_loc_df) %>% unique() %>% mutate(ID = rownames(loc_df)) %>%
#   select(chr = seqnames, start, end, ID, SYMBOL)
```

# Functional annotation of CpGs

## CpG Islands

```{r}
islands_anno <- rbind(as.data.frame(veh_cpg_anno_gr)[, c("CpG_ID", "Relation_to_Island")] %>% mutate(Treatment = "Baseline"),
                      as.data.frame(dex_cpg_anno_gr)[, c("CpG_ID", "Relation_to_Island")] %>% mutate(Treatment = "GR-stimulated")) %>% 
  group_by(Treatment, Relation_to_Island) %>% 
  tally() %>%
  group_by(Treatment) %>%
  mutate(freq = n/sum(n)) 

islands_anno$Relation_to_Island <- factor(islands_anno$Relation_to_Island, 
                                          levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
```

```{r out.width = "90%", fig.cap = "Distribution of locations of CpGs in different genomic regions in relation to CpG Island", include = T, eval = T}
plt_size <- 12

ggplot(islands_anno, aes(x = Relation_to_Island, y = freq, fill = Treatment, color = Treatment)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = c(.15, .75),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt_size),
        axis.title = element_text(size = plt_size),
        axis.text.x = element_text(angle = 0, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.5)) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```

## Gene regions 

```{r}
gene_regions_anno <- rbind(as.data.frame(veh_cpg_anno_gr)[, c("CpG_ID", "annotation")] %>% mutate(Treatment = "Baseline"),
                      as.data.frame(dex_cpg_anno_gr)[, c("CpG_ID", "annotation")] %>% mutate(Treatment = "GR-stimulated")) %>% 
  group_by(Treatment, annotation) %>% 
  tally() %>%
  group_by(Treatment) %>%
  mutate(freq = n/sum(n)) 

gene_regions_anno$annotation <- factor(gene_regions_anno$annotation, 
                                          levels = c("Promoter (2-3kb)", "Promoter (1-2kb)", "Promoter (<=1kb)",
                                                     "5' UTR", "1st Exon", "1st Intron", "Other Exon", "Other Intron",
                                                     "3' UTR", "Downstream (<=300)", "Distal Intergenic"))
```


```{r out.width = "90%", fig.cap = "Distribution of locations of CpGs in different genomic regions in relation to nearby gene", include = T, eval = T}
plt_size <- 12

ggplot(gene_regions_anno, aes(x = annotation, y = freq, fill = Treatment, color = Treatment)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt_size),
        axis.title = element_text(size = plt_size),
        axis.text.x = element_text(angle = 10, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.5)) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```

# Functional annotation of SNPs

```{r}
gene_regions_anno <- rbind(data.frame(veh_snp_anno_gr[, "annotation"], SNP = names(veh_snp_anno_gr), Treatment = "Baseline") %>% 
                             dplyr::select(SNP, annotation, Treatment),
                           data.frame(dex_snp_anno_gr[, "annotation"], SNP = names(dex_snp_anno_gr), Treatment = "GR-stimulated") %>% 
                             dplyr::select(SNP, annotation, Treatment)) %>%
  group_by(Treatment, annotation) %>% 
  tally() %>%
  group_by(Treatment) %>%
  mutate(freq = n/sum(n)) 

gene_regions_anno$annotation <- factor(gene_regions_anno$annotation, 
                                          levels = c("Promoter (2-3kb)", "Promoter (1-2kb)", "Promoter (<=1kb)",
                                                     "5' UTR", "1st Exon", "1st Intron", "Other Exon", "Other Intron",
                                                     "3' UTR", "Downstream (<=300)", "Distal Intergenic"))
```


```{r out.width = "90%", fig.cap = "Distribution of locations of SNPs in different genomic regions in relation to nearby gene", include = T, eval = T}
plt_size <- 12

ggplot(gene_regions_anno, aes(x = annotation, y = freq, fill = Treatment, color = Treatment)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt_size),
        axis.title = element_text(size = plt_size),
        axis.text.x = element_text(angle = 10, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.5)) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```

# ChromHMM states

## CpGs

```{r}
treatment    <- "dex"
dex_cpg_anno <- fread(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_cpgs_chromhmm_annotated.rds"))

treatment    <- "veh"
veh_cpg_anno <- fread(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_cpgs_chromhmm_annotated.rds"))
```


```{r}
chromhmm_anno <- rbind(unique(veh_cpg_anno[, c("ID", "annot.type")]) %>% mutate(Treatment = "Baseline"),
                       unique(dex_cpg_anno[, c("ID", "annot.type")]) %>% mutate(Treatment = "GR-stimulated")) %>% 
  group_by(Treatment, annot.type) %>% 
  tally() %>%
  group_by(Treatment) %>%
  mutate(freq = n/sum(n)) 

chromhmm_anno$annot.type <- factor(chromhmm_anno$annot.type, 
                                          levels = c("1_TssA", "2_TssAFlnk", "3_TxFlnk", "4_Tx", "5_TxWk", "6_EnhG", "7_Enh",
                                                     "8_ZNF/Rpts", "9_Het", "10_TssBiv", "11_BivFlnk", "12_EnhBiv", "13_ReprPC",
                                                     "14_ReprPCWk", "15_Quies"))
```

```{r out.width = "90%", fig.cap = "Distribution of locations of CpGs in chrom states", include = T, eval = T}
plt_size <- 14

ggplot(chromhmm_anno, aes(x = annot.type, y = freq, fill = Treatment, color = Treatment)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt_size),
        axis.title = element_text(size = plt_size),
        axis.text.x = element_text(angle = 0, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.5)) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```

## SNPs

```{r}
treatment    <- "dex"
dex_snp_anno <- fread(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_snps_chromhmm_annotated.rds"))

treatment    <- "veh"
veh_snp_anno <- fread(paste0(rslt_dir, treatment, "/smccnet_", treatment, "_snps_chromhmm_annotated.rds"))
```


```{r}
chromhmm_anno <- rbind(unique(veh_snp_anno[, c("ID", "annot.type")]) %>% mutate(Treatment = "Baseline"),
                       unique(dex_snp_anno[, c("ID", "annot.type")]) %>% mutate(Treatment = "GR-stimulated")) %>% 
  group_by(Treatment, annot.type) %>% 
  tally() %>%
  group_by(Treatment) %>%
  mutate(freq = n/sum(n)) 

chromhmm_anno$annot.type <- factor(chromhmm_anno$annot.type, 
                                          levels = c("1_TssA", "2_TssAFlnk", "3_TxFlnk", "4_Tx", "5_TxWk", "6_EnhG", "7_Enh",
                                                     "8_ZNF/Rpts", "9_Het", "10_TssBiv", "11_BivFlnk", "12_EnhBiv", "13_ReprPC",
                                                     "14_ReprPCWk", "15_Quies"))
```

```{r out.width = "90%", fig.cap = "Distribution of locations of SNPs in chrom states", include = T, eval = T}
plt_size <- 14

ggplot(chromhmm_anno, aes(x = annot.type, y = freq, fill = Treatment, color = Treatment)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(freq, accuracy = 0.1)),
            vjust = -.5,  position = position_dodge(1), size = 3, color = "black") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "", 
       title = "") +
  theme(legend.title = element_blank(), 
        legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = plt_size),
        axis.title = element_text(size = plt_size),
        axis.text.x = element_text(angle = 0, hjust = 0.5, colour = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  scale_color_manual(values = alpha(cbPalette, 0.5)) +
  scale_fill_manual(values = alpha(cbPalette, 0.25))
```


